### ISSUE TITLE
Support labels for instances in node pool

### ISSUE THREAD

**Issue by JanVoracek** [hashicorp/terraform-provider-google#1758]
### Terraform Version
```
Terraform v0.11.7
+ provider.google v1.15.0
+ provider.kubernetes v1.1.0
```

### Affected Resource(s)
Please list the resources as a list, for example:
- google_container_node_pool

### Terraform Configuration Files
```hcl
resource "google_container_node_pool" "my_pool" {
  name       = "my-pool"
  cluster    = "my-cluster"
  node_count = "2"

  node_config {
    labels {
      "foo" = "bar"
    }
  }
}
```

We can specify Kubernetes labels for nodes in `google_container_node_pool` (using `node_config.labels`) but it's not possible to set labels for the instances themselves (similarly to `google_compute_instance .labels`).

b/299443048

**Comment by dwradcliffe** [comment]
Google does not currently provide any way to do this. I have requested this feature but it has not been done yet.

**Comment by JanVoracek** [comment]
It's probably a stupid question but why it's not possible if `gcloud compute instances add-labels` can do it?

**Comment by dwradcliffe** [comment]
Since the instances are managed by the node pool (and the underlying managed instance groups), you can't make changes directly to the instances (via the method you mention). Instead we must modify the node pool itself, and there's no API for setting resource labels on the node pool. 😞

**Comment by ianrose14** [comment]
Although [the docs](https://www.terraform.io/docs/providers/google/r/container_node_pool.html) don't mention it, this seems to work for me in practice.  I'm using:
```
Terraform v0.11.8
+ provider.google v1.19.0
```
It appears to support taints as well.

**Comment by joostvdg** [comment]
@ianrose14 can you share the code you used to set labels and taints that end up on the node instances?

**Comment by ianrose14** [comment]
Here's one example
```
resource "google_container_node_pool" "name" {
  name       = "my-pool"
  cluster    = "${google_container_cluster.primary.name}"
  zone       = "${var.primary_zone}"
  node_count = 1

  autoscaling {
    min_node_count = 1
    max_node_count = 4
  }

  node_config {
    labels = {
      "cluster"                     = "${var.cluster_node_label}"
      "k8s.fullstory.com/node-task" = "my-value"
    }

    machine_type = "n1-standard-4"
    oauth_scopes = ["https://www.googleapis.com/auth/cloud-platform"]
    preemptible  = "true"

    taint = {
      effect = "NO_SCHEDULE"
      key    = "fs-node-use"
      value  = "my-value"
    }
  }
}
```

**Comment by martinasia** [comment]
If the goal is to labelled all the Nodes/VMs for that particular cluster, I normally set it up under "**google_container_cluster**" resource by defining the **resource_labels** argument. Not sure if you want to take this approach but for me this one works just like what I wanted (having custom labels for all my cluster nodes). Here's a piece of code example that I normally use :

```
resource "google_container_cluster" "primary" {
  name                          = "k8s-${terraform.workspace}-cluster"
  zone                          = "${var.region}-a"
  enable_legacy_abac            = false
  remove_default_node_pool      = true
  node_pool {
    name = "default-pool"
  }
  resource_labels               = "${var.resource_labels}"
}

# Labeling for billing purposes #
variable "resource_labels" {
  default = {
    environment = "development"
    maintainer  = "iamfiverrguy@gmail.com"
  }
  description = "Kubernetes cluster-wide resource labels"
}

**Comment by red8888** [comment]
> Here's one example
> 
> ```
> resource "google_container_node_pool" "name" {
>   name       = "my-pool"
>   cluster    = "${google_container_cluster.primary.name}"
>   zone       = "${var.primary_zone}"
>   node_count = 1
> 
>   autoscaling {
>     min_node_count = 1
>     max_node_count = 4
>   }
> 
>   node_config {
>     labels = {
>       "cluster"                     = "${var.cluster_node_label}"
>       "k8s.fullstory.com/node-task" = "my-value"
>     }
> 
>     machine_type = "n1-standard-4"
>     oauth_scopes = ["https://www.googleapis.com/auth/cloud-platform"]
>     preemptible  = "true"
> 
>     taint = {
>       effect = "NO_SCHEDULE"
>       key    = "fs-node-use"
>       value  = "my-value"
>     }
>   }
> }
> ```

This can probably be closed right? labels under node_config works.
applying resource labels at google_container_cluster does NOT work- it will not apply those labels to the GCE instances

**Comment by martinasia** [comment]
Well... I tested this just a while ago and I must say that adding resource_labels argument to  google_container_cluster DOES work for me. I was also skeptical at first and hesitated when I tried this method last year (although the TF docs which I read says it very clearly) here: [https://www.terraform.io/docs/providers/google/r/container_cluster.html#resource_labels](url) , however, I went ahead anyway and gave it a test.
Long story to short, I'm glad that I took that route since I eventually learned something about how google_container_cluster and that particular argument (resource_labels) works.

The very thing which I learned was, the (cluster-wide) label changes are not instantaneous! It took about roughly 30~45mins for me to see the changes propagated and reflected on ALL our GCE instances (nodes). If you have several node pools and you want to make one time shot of applying labels to the pools then this will make things easier for you (assuming that you have that level of patience required).

So...perhaps you could give this another try (by sparing some extra time to get the changes fully reflected across all the nodes) and share the result here @red8888 ? I'm sure your experience in testing this technique will benefit us all.

I also post some of the cli stdouts which I performed during the test where I added 'maintainer=Martin' as a resource_labels to my TF' vars file:

```
$ terraform -v
Terraform v0.12.1
+ provider.google v2.7.0
```
```
$ terraform plan
Terraform will perform the following actions:

  # module.gke.google_container_cluster.primary will be updated in-place
  ~ resource "google_container_cluster" "primary" {
        additional_zones         = []
        cluster_autoscaling      = []
        cluster_ipv4_cidr        = "10.20.0.0/14"
        ...
        master_version           = "1.12.7-gke.17"
        min_master_version       = "1.12.7-gke.17"
        monitoring_service       = "monitoring.googleapis.com"
        name                     = "k8s-dev-cluster"
        node_locations           = []
        node_version             = "1.12.7-gke.17"
        project                  = "xxx"
        remove_default_node_pool = true
      ~ resource_labels          = {
            "env"        = "staging"
          + "maintainer" = "martin"
            "resource"   = "gke"
        }
        ...
```
After roughly about an hour, I checked one of the nodes (GCE) for that cluster by issuing the following command:
``` 
$ gcloud compute instances describe <one of your nodes> --zone=<the gke zone> | head --lines=+50

canIpForward: true
cpuPlatform: Intel Broadwell
creationTimestamp: '2019-06-11T05:45:50.194-07:00'
deletionProtection: false
disks:
...
id: 'xxx'
kind: compute#instance
labelFingerprint: ptUBrXARDBY=
labels:
  env: staging
  goog-gke-node: ''
  maintainer: martin    <<<<<<<<<<
  resource: gke
...
```

Lastly, I've tested this and the resource argument worked as it should on both preemptible and non-preemptible instance type.

**Comment by talonx** [comment]
I'm curious to know why this is not supported yet with google_container_node_pool when the other similar module https://registry.terraform.io/modules/terraform-google-modules/kubernetes-engine/google/9.1.0 supports this. 

E.g. see the example at https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/tree/v9.1.0/modules/beta-public-cluster

I would rather use the resource google_container_node_pool directly than the other module, because this is much more flexible.

**Comment by rileykarson** [comment]
@talonx: Those are Kubernetes labels as specified with `node_config.labels` and not GCP labels, which are not currently able to be set on node pools.

**Comment by rileykarson** [comment]
`resource_labels`, the cluster-level field, applies its labels to the GCE resources managed by the cluster. At the very least, it does for the nodes (i.e. GCE VMs) and I'm not sure offhand if it does for Kubernetes-managed resources like load balancers.

The ask here as we've interpreted it is for that kind of management at the node pool level, rather than across the whole cluster- that way, two different pools could apply different sets of labels to their nodes. That would make a config like the following possible:

```tf
resource "google_container_cluster" "primary" {
  name     = "my-gke-cluster"
  location = "us-central1"

  resource_labels {
    cluster-label = "some-value"
  }

  remove_default_node_pool = true
  initial_node_count       = 1
}

resource "google_container_node_pool" "preemptible_nodes" {
  name       = "my-node-pool-preemptible"
  cluster    = google_container_cluster.primary.id
  node_count = 1

  node_config {
    preemptible  = true
    machine_type = "e2-medium"
    resource_labels {
      preemptible-node-label = "some-other-value"
    }
  }
}

resource "google_container_node_pool" "primary_nodes" {
  name       = "my-node-pool"
  cluster    = google_container_cluster.primary.id
  node_count = 1

  node_config {
    machine_type = "e2-large"
    resource_labels {
      primary-node-label = "some-distinct-value"
    }
  }
}
```

**Comment by carlos4ndre** [comment]
@rileykarson It would be great to have support for `resource_labels` at the node pool level, as you highlighted in your example.

Looking at the [REST API](https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1/NodeConfig) and [SDK](https://cloud.google.com/sdk/gcloud/reference/container/node-pools/create) this feature doesn't seem to be supported yet, so at the moment the default behaviour is to set the cluster-wide `resource_labels` in all instance templates used by the node pools.

For reference, believe this feature is supported in [EKS](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/eks_node_group#tags) and [AKS](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster_node_pool#tags).

**Comment by carlos4ndre** [comment]
Having support for `resource_labels` at the node pool level is particularly useful for billing, as we can do cost allocation per Team by setting a `team` label.

**Comment by korenlev** [comment]
If the node_pool is configured with auto scaling enabled then Update any metadata in node pool will FAIL with the following google error:

"Updates for 'labels' are not supported in node pools with autoscaling enabled (as a workaround, consider temporarily disabling autoscaling or recreating the node pool with the updated values.)"

You can publish this error back to the user,
so he can decide to use this temp workaround by google or not.
You can validate this with upstream (the relationship with auto scaling property).

**Comment by hudac** [comment]
Looking into the [REST API](https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1/NodeConfig), applying labels to a GCP node pool resource seems to be supported now:
```
  "resourceLabels": {
    string: string,
    ...
  },
```

Tried with the `hashicorp/google v5.15.0` provider, adding `node_config.resource_labels` works:

```
$ terraform apply
...

  ~ resource "google_container_node_pool" "np" {

      ~ node_config {
          ~ resource_labels   = {
              + "team" = "dev"
              + "unit" = "rnd"
            }
            # (14 unchanged attributes hidden)

            # (2 unchanged blocks hidden)
        }
```

**Comment by rileykarson** [comment]
Ah, yep, seems this was resolved in https://github.com/GoogleCloudPlatform/magic-modules/pull/6842 and released in `4.45.0`
